src/replay/service.zig

```diff
@@ -91,19 +96,1269 @@ fn advanceReplay(state: *ReplayState) !void {
+        // The original `SlotStateUpdate::DuplicateConfirmed` is a no-op when the bank has not been
+        // replayed yet, and unlike 2) there is no upcoming `SlotStateUpdate::BankFrozen` or
+        // `SlotStateUpdate::Dead`, as `slot` is pruned and will not be replayed.
+        //
+        // Thus if we have a duplicate confirmation, but `slot` is pruned, we continue
+        // processing it as `epoch_slots_frozen`.
+        if (!is_popular_pruned) {
+            if (maybe_duplicate_confirmed_hash) |duplicate_confirmed_hash| {
+                if (!epoch_slots_frozen_hash.eql(duplicate_confirmed_hash)) {
+                    logger.warn().logf(
+                        "EpochSlots sample returned slot {} with hash {}, " ++
+                            "but we already saw duplicate confirmation on hash: {}",
+                        .{ slot, epoch_slots_frozen_hash, duplicate_confirmed_hash },
+                    );
+                }
+                return;
+            }
+        }
+
+        switch (bank_status) {
+            .frozen => |bank_frozen_hash| {
+                if (bank_frozen_hash.eql(epoch_slots_frozen_hash)) {
+                    // Matches, nothing to do
+                    return;
+                } else {
+                    // The epoch slots hash does not match our frozen hash.
+                    logger.warn().logf(
+                        "EpochSlots sample returned slot {} with hash {}, " ++
+                            "but our version has hash {}",
+                        .{ slot, epoch_slots_frozen_hash, bank_frozen_hash },
+                    );
+                    if (!is_popular_pruned) {
+                        // If the slot is not already pruned notify fork choice to mark as invalid
+                        try resulting_change.markSlotDuplicate(slot, fork_choice, bank_frozen_hash);
+                    }
+                }
+            },
+            .dead => {
+                // Cluster sample found a hash for our dead slot, we must have the wrong version
+                logger.warn().logf(
+                    "EpochSlots sample returned slot {} with hash {}, " ++
+                        "but we marked slot dead",
+                    .{ slot, epoch_slots_frozen_hash },
+                );
+            },
+            .unprocessed => {
+                // If the bank was not popular pruned, we would never have made it here, as the bank is
+                // yet to be replayed
+                std.debug.assert(is_popular_pruned);
+                // The cluster sample found the troublesome slot which caused this fork to be pruned
+                logger.warn().logf(
+                    "EpochSlots sample returned slot {} with hash {}, " ++
+                        "but we have pruned it due to incorrect ancestry",
+                    .{ slot, epoch_slots_frozen_hash },
+                );
+            },
+        }
+
+        try resulting_change.repairDuplicateConfirmedVersion(
```

Comment:
nit: Any reason not to have it as `duplicate_slots_to_repair.put(allocator, slot, epoch_slots_frozen_hash)`

I.e remove the need for the static function?